pkg/compact/dedup/merger.go

```diff
@@ -0,0 +1,506 @@
+package dedup
+
+import (
+	"context"
+	"fmt"
+	"math/rand"
+	"os"
+	"path/filepath"
+	"sort"
+	"strings"
+	"sync"
+	"time"
+
+	"github.com/go-kit/kit/log"
+	"github.com/go-kit/kit/log/level"
+	"github.com/oklog/ulid"
+	"github.com/pkg/errors"
+	"github.com/prometheus/tsdb/chunks"
+	"github.com/prometheus/tsdb/labels"
+	"github.com/thanos-io/thanos/pkg/block"
+	"github.com/thanos-io/thanos/pkg/block/metadata"
+	"github.com/thanos-io/thanos/pkg/compact"
+	"github.com/thanos-io/thanos/pkg/compact/downsample"
+	"github.com/thanos-io/thanos/pkg/objstore"
+	"github.com/thanos-io/thanos/pkg/query"
+)
+
+type TimeWindow struct {
+	MinTime int64
+	MaxTime int64
+}
+
+func (tw *TimeWindow) String() string {
+	return fmt.Sprintf("[%d, %d]", tw.MinTime, tw.MaxTime)
+}
+
+func NewTimeWindow(minTime, maxTime int64) *TimeWindow {
+	return &TimeWindow{MinTime: minTime, MaxTime: maxTime}
+}
+
+// Group blocks under the same time window from different replicas
+type BlockGroup struct {
+	window *TimeWindow
+	blocks []*metadata.Meta
+}
+
+func (g *BlockGroup) String() string {
+	builder := strings.Builder{}
+	builder.WriteString("[")
+	for i, b := range g.blocks {
+		if i != 0 {
+			builder.WriteString(",")
+		}
+		builder.WriteString(b.ULID.String())
+	}
+	builder.WriteString("]")
+	return fmt.Sprintf("BlockGroup{window: %s, blocks: %s}", g.window, builder.String())
+}
+
+func NewBlockGroup(window *TimeWindow, blocks []*metadata.Meta) *BlockGroup {
+	return &BlockGroup{window: window, blocks: blocks}
+}
+
+type BlockGroups []*BlockGroup
+
+func NewBlockGroups(replicas Replicas) BlockGroups {
```

Comment:
This might be doable in Iterator as well. I think it would more understandable if we would have just one method/algo for this - for both offline and online deduplication. Online deduplication is available here: https://github.com/thanos-io/thanos/blob/master/pkg/query/iter.go#L304

Reply:
The `BlockGroups ` provides the function to group all the blocks under the specified time range. It doesn't provide any deduplication here.